pkg/controller/machine/machine_controller.go

```diff
@@ -1065,16 +1065,21 @@ func (r *Reconciler) getNode(ctx context.Context, instance instance.Instance, pr
 		return nil, false, err
 	}
 
-	// We trim leading slashes in raw ID, since we always want three slashes in full ID
-	providerID := fmt.Sprintf("%s:///%s", provider, strings.TrimLeft(instance.ID(), "/"))
 	for _, node := range nodes.Items {
 		if provider == providerconfigtypes.CloudProviderAzure {
 			// Azure IDs are case-insensitive
+			// We trim leading slashes in raw ID, since we always want three slashes in full ID
+			providerID := fmt.Sprintf("%s:///%s", provider, strings.TrimLeft(instance.ID(), "/"))
 			if strings.EqualFold(node.Spec.ProviderID, providerID) {
 				return node.DeepCopy(), true, nil
 			}
 		} else {
-			if node.Spec.ProviderID == providerID {
+			// We trim leading slashes in raw ID, since we always want three slashes in full ID
+			// Some provider IDs have three slashes (e.g. AWS), but some have two (e.g. GCP, vSphere)
+			// To mitigate this, we split on "://", and then compare provider part (the first part),
+			// and the ID part (the second part) specially
+			pid := strings.Split(node.Spec.ProviderID, "://")
+			if len(pid) == 2 && pid[0] == string(provider) && strings.TrimLeft(pid[1], "/") == strings.TrimLeft(instance.ID(), "/") {
```

Comment:
nit: consider moving to function and adding some unittest with some realistic cloud provider ID examples.

Reply:
Added some unit tests, however, it's hard to test ProviderID generation for real providers because we don't have unit tests for provider implementations.